## FINAL ASSIGNMENT

In [34]:
import pandas as pd

In [35]:
df = pd.read_csv(r"C:\Users\HP i5\Desktop\SE_Factory\Week 1\Pythonex\kc_house_data.csv")

In [36]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [38]:
df['date'] = pd.to_datetime(df['date'])

To start with the analysis, i began with comparing the attributes and features of the most expensive houses with those of the cheapest houses. in order to get some initial insights on what to focus on through the analysis of this dataset.

In [39]:
df.nlargest(5, 'price')

id       date      price  bedrooms  bathrooms  sqft_living  \
7252  6762700020 2014-10-13  7700000.0         6       8.00        12050   
3914  9808700762 2014-06-11  7060000.0         5       4.50        10040   
9254  9208900037 2014-09-19  6890000.0         6       7.75         9890   
4411  2470100110 2014-08-04  5570000.0         5       5.75         9200   
1448  8907500070 2015-04-13  5350000.0         5       5.00         8000   

      sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
7252     27600     2.5           0     3  ...     13        8570   
3914     37325     2.0           1     2  ...     11        7680   
9254     31374     2.0           0     4  ...     13        8860   
4411     35069     2.0           0     0  ...     13        6200   
1448     23985     2.0           0     4  ...     12        6720   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
7252           3480      1910          1987    98102  47.6298 -122.323   
3914           2360      1940          2001    98004  47.6500 -122.214   
9254           1030      2001             0    98039  47.6305 -122.240   
4411           3000      2001             0    98039  47.6289 -122.233   
1448           1280      2009             0    98004  47.6232 -122.220   

      sqft_living15  sqft_lot15  
7252           3940        8800  
3914           3930       25449  
9254           4540       42730  
4411           3560       24345  
1448           4600       21750  

[5 rows x 21 columns]

The subsets with the highest prices (Above 7 million dollars) have various number of bedrooms/bathrooms, high living area, good view levels, higher grades (above 11) significantly higher basement area. All of these may contribute to the pricing of these properties, which is often true as more expensive houses tend to have luxurious features.  

In [40]:
df.nsmallest(5, 'price')

id       date    price  bedrooms  bathrooms  sqft_living  \
1149   3421079032 2015-02-17  75000.0         1       0.00          670   
15293    40000362 2014-05-06  78000.0         2       1.00          780   
465    8658300340 2014-05-23  80000.0         1       0.75          430   
16198  3028200080 2015-03-24  81000.0         2       1.00          730   
8274   3883800011 2014-11-05  82000.0         3       1.00          860   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
1149      43377     1.0           0     0  ...      3         670   
15293     16344     1.0           0     0  ...      5         780   
465        5050     1.0           0     0  ...      4         430   
16198      9975     1.0           0     0  ...      5         730   
8274      10426     1.0           0     0  ...      6         860   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
1149               0      1966             0    98022  47.2638 -121.906   
15293              0      1942             0    98168  47.4739 -122.280   
465                0      1912             0    98014  47.6499 -121.909   
16198              0      1943             0    98168  47.4808 -122.315   
8274               0      1954             0    98146  47.4987 -122.341   

       sqft_living15  sqft_lot15  
1149            1160       42882  
15293           1700       10387  
465             1200        7500  
16198            860        9000  
8274            1140       11250  

[5 rows x 21 columns]

The subsets with the cheapest houses(Around 80,000$) have the very low area (less than one thousand) smaller number of bedrooms and bathrooms, only one floor, 0 waterfront, no view, low grade leveks (less than 7), and no basement. Those features may be influecing the pricing of the properties.  

After doing an initial analysis on the prices, i will apply the correlation function to get some additional insights on the relationship between each attribute and the target(price). The correlation function gives a values between 1(Strong positive correlation) to -1(Strong negative correlation) with 0 correlation meaning no relationship. It reflects how similar the measurements of variables are across a dataset and identifies patterns and relationships that exist within the data. However, the default .corr() function (which uses pearson's method) is usually applied on numeric variables.  

In [78]:
correlation = df.corr()
price_corr = correlation['price'].sort_values(ascending=False)
print(price_corr)

price            1.000000
sqft_living      0.702044
grade            0.667463
sqft_above       0.605566
sqft_living15    0.585374
bathrooms        0.525134
view             0.397346
sqft_basement    0.323837
bedrooms         0.308338
lat              0.306919
flag_bedroom     0.296029
waterfront       0.266331
floors           0.256786
yr_renovated     0.126442
sqft_lot         0.089655
sqft_lot15       0.082456
yr_built         0.053982
condition        0.036392
long             0.021571
date            -0.004366
id              -0.016797
zipcode         -0.053168
Name: price, dtype: float64


The correlation method provides us with some good insights; there is a strong correlation between the price and the square foot of living space. However concerning the different variable we have; categorical variables and geographical variables, there isn't much to inspect through this method and it doesnt provide good values on such correlation. The df.corr() method works on continuous numerical variables, but not for ranked/geographic variables. The "kendall" method on the other side, does work on ranked variables and gives better results for them.

In [42]:
correlation = df.corr(method='kendall')
price_corr = correlation['price'].sort_values(ascending=False)
print(price_corr)

price            1.000000
grade            0.527844
sqft_living      0.463620
sqft_living15    0.405037
sqft_above       0.378244
bathrooms        0.369216
lat              0.292903
bedrooms         0.265100
floors           0.251630
view             0.237574
sqft_basement    0.190399
waterfront       0.094071
yr_renovated     0.082378
yr_built         0.070321
sqft_lot         0.047191
long             0.039866
sqft_lot15       0.038425
condition        0.014485
id               0.003218
date            -0.007266
zipcode         -0.009826
Name: price, dtype: float64


Through this kendall correlation method, the attribute with the highest correlation is the grade, hence we have to focus on it through our analysis to get further insights about it.

After these initial observations on the dataset, and the correlations between its various attributes, I will focus on them to go further through my analysis and get better results about each of them.

*Bedrooms and bathroom:* 

In [43]:
df['bedrooms'].value_counts()

bedrooms
3     9824
4     6882
2     2760
5     1601
6      272
1      199
7       38
0       13
8       13
9        6
10       3
11       1
33       1
Name: count, dtype: int64

More than 50% of the houses are with 3 or less bedrooms. i will be dividing the houses into two subsets, to take a look at each alone and analyze it more clearly. One of the subsets is going to be houses with 3 bedrooms and less, and the other will be of 4 bedrooms and more. 

In [44]:
df['flag_bedroom'] = df.apply(lambda x: 1 if x['bedrooms'] >= 4 else 0, axis=1)

In [71]:
df.groupby('flag_bedroom').agg({'id':['count'], 'price': ['min', 'max', 'mean'], 'sqft_living':['mean'], 
                                    'grade':['median'] })

id     price                            sqft_living  grade
              count       min        max           mean         mean median
flag_bedroom                                                               
0             12796   75000.0  3800000.0  449912.335495  1669.310957    7.0
1              8817  100000.0  7700000.0  671189.605648  2675.782012    8.0

In [74]:
df.loc[df['bedrooms']==0, ['price', 'bedrooms', 'bathrooms', 'sqft_living']].reset_index()


index      price  bedrooms  bathrooms  sqft_living
0     875  1100000.0         0       0.00         3064
1    3119   380000.0         0       0.00         1470
2    3467   288000.0         0       1.50         1430
3    4868   228000.0         0       1.00          390
4    6994  1300000.0         0       0.00         4810
5    8477   339950.0         0       2.50         2290
6    8484   240000.0         0       2.50         1810
7    9773   355000.0         0       0.00         2460
8    9854   235000.0         0       0.00         1470
9   12653   320000.0         0       2.50         1490
10  14423   139950.0         0       0.00          844
11  18379   265000.0         0       0.75          384
12  19452   142000.0         0       0.00          290

In [70]:
df.loc[df['bedrooms']==33, ['price', 'bathrooms', 'sqft_living']]



price  bathrooms  sqft_living
15870  640000.0       1.75         1620

The dataframes show that houses with bedrooms 4 and more are usually of higher area and higher prices. As shown in the table, some of the houses have zero bedrooms but different number of bathrooms, meaning that they are either a studio house or there is a mistake in the data implemented. Also there is a house with 33 bedrooms however a rather small area and cheap price, so this must be a typo. 

In [48]:
df.loc[(df['bedrooms'] == 8) & (df['bathrooms'] == 2.75) ]

id       date      price  bedrooms  bathrooms  sqft_living  \
6174   1873400020 2014-07-03   340000.0         8       2.75         2790   
10958  1652500060 2014-07-11  1650000.0         8       2.75         4040   
15670  8141200080 2014-08-14   680000.0         8       2.75         2530   

       sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
6174       6695     1.0           0     0  ...        1470           1320   
10958     20666     1.0           0     0  ...        2020           2020   
15670      4800     2.0           0     0  ...        1390           1140   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
6174       1977             0    98133  47.7565 -122.331           1760   
10958      1962             0    98004  47.6340 -122.221           3670   
15670      1901             0    98112  47.6241 -122.305           1540   

       sqft_lot15  flag_bedroom  
6174         7624             1  
10958       20500             1  
15670        4800             1  

[3 rows x 22 columns]

For equal numbers of bedrooms and bathrooms, we have huge differences in the prices, for apartments of id "1873400020" and "8141200080", the price is doubled even though we have equal number of bedrooms, bathrooms, and very close number of sqaure feet of living space. This doubling of price may be due to "increment a level in floor" or "difference in latitudes and longtitudes hence being in a different area" as all of other factors doesn't differ much.
Also, the 10x increase of price of the "1652500060" house is influenced by square feet of living space and square feet of basement, also this house is in a different zipcode. All those would have increased the price of this house by this much. 

In [49]:
correlation = (df['bathrooms'] + df['bedrooms']).corr(df['price'])
print(correlation)


0.46631014225336537


In [50]:
correlation = (df['bathrooms'] + df['bedrooms'] + df['sqft_living']).corr(df['price'])
print(correlation)

0.7019418010114713


What is noticed that, an increase in the square feet of living space has a high influence, reassuring what was indicated by the correlation function that the sqft_living is of the highest influence on price. 

*Square footage of living space*

In [85]:
df['sqft_living2'] = df.apply(lambda x: 'High' if x['sqft_living'] >= 10000
                     else 'Medium' if x['sqft_living'] > 5000 
                     else 'Low' if x['sqft_living'] >1000 
                     else 'very low' , axis=1)
df.head()

id       date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520 2014-10-13  221900.0         3       1.00         1180   
1  6414100192 2014-12-09  538000.0         3       2.25         2570   
2  5631500400 2015-02-25  180000.0         2       1.00          770   
3  2487200875 2014-12-09  604000.0         4       3.00         1960   
4  1954400510 2015-02-18  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_basement  yr_built  \
0      5650     1.0           0     0  ...              0      1955   
1      7242     2.0           0     0  ...            400      1951   
2     10000     1.0           0     0  ...              0      1933   
3      5000     1.0           0     0  ...            910      1965   
4      8080     1.0           0     0  ...              0      1987   

   yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  \
0             0    98178  47.5112 -122.257           1340        5650   
1          1991    98125  47.7210 -122.319           1690        7639   
2             0    98028  47.7379 -122.233           2720        8062   
3             0    98136  47.5208 -122.393           1360        5000   
4             0    98074  47.6168 -122.045           1800        7503   

   flag_bedroom  sqft_living2  
0             0           Low  
1             0           Low  
2             0      very low  
3             1           Low  
4             0           Low  

[5 rows x 23 columns]

In [87]:
df[['price', 'bedrooms', 'bathrooms', 'sqft_living2']].groupby(['sqft_living2']).agg({'price':['min', 'max', 'mean'], 
                                                                                    'bedrooms':['min', 'max', 'median'], 
                                                                                    'bathrooms': ['min', 'max', 'median']})

price                          bedrooms             \
                    min        max          mean      min max median   
sqft_living2                                                           
High          2280000.0  7700000.0  5.680000e+06        5   7    6.0   
Low             95000.0  3640000.0  5.421001e+05        0  33    3.0   
Medium         452000.0  6890000.0  2.019841e+06        2   8    5.0   
very low        75000.0   747000.0  2.983010e+05        0   4    2.0   

             bathrooms               
                   min   max median  
sqft_living2                         
High              4.50  8.00  8.000  
Low               0.00  7.50  2.250  
Medium            1.75  7.75  4.125  
very low          0.00  2.75  1.000

Through the above dataframe, as the square footage of living space increase, the price tends to increase as well. This shows a strong positive correlation between the square footage area and the price. Also, as the sqft living increase the number of bedrooms and bathrooms increase. Where it is typical that larger homes have more bedrooms and bathrooms.

In [94]:
df[['sqft_living2', 'yr_built', 'yr_renovated', 'price']].groupby('sqft_living2').agg({'yr_built':['count'], 'yr_renovated':['count'], 
                                                                                        'price' : ['min', 'max', 'mean']})

yr_built yr_renovated      price                         
                count        count        min        max          mean
sqft_living2                                                          
High                3            3  2280000.0  7700000.0  5.680000e+06
Low             19908        19908    95000.0  3640000.0  5.421001e+05
Medium            208          208   452000.0  6890000.0  2.019841e+06
very low         1494         1494    75000.0   747000.0  2.983010e+05

*Grade*

In [97]:
df[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'grade', 'floors']].groupby(['grade']).agg({
                                                                                'price' : ['min', 'max', 'mean'], 
                                                                                'sqft_living' : ['min', 'max', 'mean'], 
                                                                                'bedrooms' : 'median', 'bathrooms' : 'median', 
                                                                                'floors' : 'median'
                                                                                })

price                          sqft_living                      \
             min        max          mean         min    max         mean   
grade                                                                       
1       142000.0   142000.0  1.420000e+05         290    290   290.000000   
3        75000.0   280000.0  2.056667e+05         520    670   596.666667   
4        80000.0   435000.0  2.143810e+05         384   1470   660.482759   
5        78000.0   795000.0  2.485240e+05         370   2040   983.326446   
6        82000.0  1200000.0  3.019166e+05         390   3900  1191.561335   
7        90000.0  2050000.0  4.025933e+05         550   4480  1689.400401   
8       140000.0  3070000.0  5.428955e+05         750   5370  2184.748517   
9       230000.0  2700000.0  7.737382e+05         860   6900  2868.139962   
10      316000.0  3600000.0  1.072347e+06        1180   6630  3520.299824   
11      420000.0  7060000.0  1.497792e+06        2370  10040  4395.448622   
12      835000.0  5350000.0  2.192500e+06        3690  13540  5471.588889   
13     1780000.0  7700000.0  3.710769e+06        4500  12050  7483.076923   

      bedrooms bathrooms floors  
        median    median median  
grade                            
1          0.0     0.000    1.0  
3          1.0     0.000    1.0  
4          2.0     1.000    1.0  
5          2.0     1.000    1.0  
6          3.0     1.000    1.0  
7          3.0     1.750    1.0  
8          3.0     2.500    2.0  
9          4.0     2.500    2.0  
10         4.0     2.750    2.0  
11         4.0     3.500    2.0  
12         4.0     3.625    2.0  
13         5.0     5.000    2.0

There is a proportional relationship between the grade and the price. As the grade increases so does the average price per each grade. Hence a higher grades properties are associated with higher prices. Also, higher grades tend to have more sqft living, bathrooms and bedrooms, and more number of floors, which would imply that higher grades have larger and more luxurious houses. This aligns with the result of the correlation function, where the "grade" attribute had the second highest influence on the price. There is also a strong positive correlation between the grade and the other attributes.

*Floors*

In [101]:
df.floors.value_counts(normalize=True).mul(100)

floors
1.0    49.414704
2.0    38.129829
1.5     8.837274
3.0     2.836256
2.5     0.744922
3.5     0.037015
Name: proportion, dtype: float64

In [103]:
df[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'grade', 'view', 'floors']].groupby(['floors']).agg({
                                                                                'price' : ['min', 'max', 'mean'], 
                                                                                'sqft_living' : ['min', 'max', 'mean'], 
                                                                                'bedrooms' : 'median', 'bathrooms' : 'median', 
                                                                                'grade' : 'median'
                                                                                }).reset_index()

floors     price                          sqft_living                      \
               min        max          mean         min    max         mean   
0    1.0   75000.0  4670000.0  4.422196e+05         290   9640  1731.846161   
1    1.5   92000.0  3650000.0  5.590449e+05         490   5530  1907.880628   
2    2.0   90000.0  7060000.0  6.490515e+05         550  10040  2571.279092   
3    2.5  255000.0  7700000.0  1.061021e+06         990  12050  3145.428571   
4    3.0  205000.0  3100000.0  5.826201e+05         760  13540  1788.311582   
5    3.5  435000.0  3300000.0  9.339375e+05        1310   7710  2518.000000   

  bedrooms bathrooms  grade  
    median    median median  
0      3.0     1.750    7.0  
1      3.0     1.750    7.0  
2      4.0     2.500    8.0  
3      4.0     2.500    9.0  
4      3.0     2.500    8.0  
5      3.0     2.625    8.0

About 50% of the houses are of 1 floor, 40% are with 2 floors. As the number of floors per house increase, so does the price except for the floor level that had a decline in price and in sqft area surprisingly. It is noticed that houses with higher floors have higher square feet are, more bedrooms and bathrooms, and are situated at a higher grade. 


*View*

In [105]:
df.view.value_counts(normalize=True).mul(100)

view
0    90.172581
2     4.455652
3     2.359691
1     1.536113
4     1.475964
Name: proportion, dtype: float64

In [107]:
df[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'grade', 'floors', 'view']].groupby(['view']).agg({
                                                                                'price' : ['mean'], 
                                                                                'sqft_living' : [ 'mean'], 
                                                                                'bedrooms' : 'median', 'bathrooms' : 'median', 
                                                                                'floors' : 'median'
                                                                                }).reset_index()

view         price  sqft_living bedrooms bathrooms floors
                mean         mean   median    median median
0    0  4.966235e+05  1997.761660      3.0      2.25    1.5
1    1  8.125186e+05  2568.960843      4.0      2.25    1.0
2    2  7.927462e+05  2655.257529      3.0      2.50    1.5
3    3  9.724684e+05  3018.564706      4.0      2.50    1.5
4    4  1.464363e+06  3351.473354      4.0      2.50    1.5

As the view level increase, so does the price. Meaning that properties of higher view levels (better views) often have higher prices. Hence, view levels play a significant role in the desirability of the appartments. People would pay more for appartments with better view, larger area, and better quality. The average price for an excellet view (view = 4) is multiplied by ten of that with no view (view = 0). This is usually the case because more luxurious houses often have better veiws along with better overall qualities. 

We can deduce from this analysis that one factor may not necessarily affect the target at its own, however it may do so along with various factors, that together may give an increased influence on the target. That's why the analysis has to be on a larger dataset and including more attributes to obtain the relationships between them

*Waterfront*

In [55]:
df[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'grade', 'view', 'waterfront']].groupby(['waterfront']).mean().round().reset_index()

waterfront      price  bedrooms  bathrooms  sqft_living  grade  view
0           0   531653.0       3.0        2.0       2072.0    8.0   0.0
1           1  1662524.0       3.0        3.0       3174.0    9.0   4.0

From this table we can conclude that the houses with waterfront often have higher prices than those without. Waterfront may be a desirable feature hence commanding a higher price. Eventhough there isn't much difference in the average number of bedrooms and bathrooms, sqft_living space, and grade (i.e. there isn't a huge relationship between these attributes and whether there is a waterfront or not). But they still may be impacting slightly. The view level increased significantly from 0 (with no waterfront) to 4 (with waterfront) explaining the huge increase in price as well. 

In [56]:
df[['price', 'waterfront', 'view', 'sqft_living']].groupby('waterfront')['view'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

waterfront  view  percentage
0           0     0   90.857809
1           0     2    4.452214
2           0     3    2.289044
3           0     1    1.543124
4           0     4    0.857809
5           1     4   82.822086
6           1     3   11.656442
7           1     2    4.907975
8           1     1    0.613497

The dataframe above shows that above 90% of the houses with no waterfront don't have a view, and a very small percentage of them (8%) have a view varying between (2-3-1-4). For houses with waterfront, however, the majority of them (83%) have an excellent view (view = 4), 11% have a verygood view, and the rest range between 2 and 1. None of the houses with waterfront is without a view. This indicates a very strong positive relationship between the waterfront and the view. This is expected as these two features are more of a luxurious features that require higher prices

*Basement area*

In [125]:
df['sqft_basement2'] = df.apply(lambda x: 0 if x['sqft_basement'] ==0 else 1 if x['sqft_basement'] <= 500 else 2 , axis=1)
df[['price', 'sqft_living', 'sqft_basement2']].groupby('sqft_basement2').mean().round(2).reset_index()

sqft_basement2      price  sqft_living
0               0  486945.39      1928.88
1               1  492036.68      1674.89
2               2  682507.34      2607.05

In [126]:
df['sqft_basement2'].value_counts()

sqft_basement2
0    13126
2     5814
1     2673
Name: count, dtype: int64

More than 50% of the houses don't have any basement, and the basement itself may have little impact on the price, we must consider other factors.

In [128]:
df['condition'].value_counts(normalize=True).mul(100)

condition
3    64.919262
4    26.275852
5     7.870263
2     0.795817
1     0.138805
Name: proportion, dtype: float64

65% of the houses are of good condition (=3), 35% of the houses are between very good and excellent conditions, and only 1% of the houses are of poor to bad conditions. This is a good scale as most houses to be sold have to be in good conditions. 

In [129]:
df.groupby('condition').agg({'price': 'mean', 'id' : 'count'})

price     id
condition                      
1          334431.666667     30
2          327316.215116    172
3          542097.086024  14031
4          521300.705230   5679
5          612577.742504   1701

The price of the houses increase as its condition is better. which is always true as better conditions demand higher prices.

*Year built and year renovated*

In [159]:
df.loc[(df['yr_built'] <2000 ) & (df['yr_renovated'] == 0), ['yr_built', 'yr_renovated', 'price', 'sqft_living'] ].sort_values(by='price', ascending=False)

yr_built  yr_renovated      price  sqft_living
1315       1991             0  5300000.0         7390
1164       1999             0  5110000.0         8010
2085       1980             0  3850000.0         5770
7035       1978             0  3800000.0         7050
19017      1997             0  3800000.0         5510
...         ...           ...        ...          ...
8274       1954             0    82000.0          860
16198      1943             0    81000.0          730
465        1912             0    80000.0          430
15293      1942             0    78000.0          780
1149       1966             0    75000.0          670

[15941 rows x 4 columns]

In [160]:
df.loc[(df['yr_built'] <2000 ) & (df['yr_renovated'] > 0), ['yr_built', 'yr_renovated', 'price', 'sqft_living']].sort_values(by='price', ascending=False)

yr_built  yr_renovated      price  sqft_living
7252       1910          1987  7700000.0        12050
3914       1940          2001  7060000.0        10040
8092       1983          2009  4670000.0         9640
2864       1941          2002  3600000.0         5020
10446      1939          2010  3400000.0         4260
...         ...           ...        ...          ...
19417      1937          1946   152275.0         1020
12984      1969          2009   146000.0          900
15540      1943          1989   135000.0          600
2589       1922          2003   134000.0          980
20962      1968          2007   110000.0          828

[913 rows x 4 columns]

Through the above dataframes, we can get many insights. 
First of all, for houses that were built before 2000 and weren't renovated, the highest prices were for the houses was were built near 2000, and the early constructed houses had much less prices. Meaning that for those houses without renovation, the newer ones had higher prices. 
Second, concerning the houses that were built before 2000 but were renovated, and despite the year they were renovated, they did have much higher prices (reaching 7 million dollars), also they had larger area. The renovated houses tend to have larger prices than those that weren't renovated. Hence the act of renovation has a positive influence on the price.
It appears that the year the house was built alone doesn't have a strong relationship with the price, however when considering the year the house was built along with the year it was renovated, the correlation with the price becomes more significant.   

In [161]:
df.loc[(df['yr_built'] >2000 ) & (df['yr_renovated'] > 0), ['yr_built', 'yr_renovated', 'price', 'sqft_living']].sort_values(by='price', ascending=False)

yr_built  yr_renovated     price  sqft_living
18185      2003          2009  825000.0         3620

Of the houses that were built after 2000 only one was renovated, which is logical because they have been built lately and don't need much modifications. (As the data was collected on 2014, hence not much of a time difference.)

The year the house was built or renovated doesn't tell much on its price. The price is affected by other factors more than these two, some houses may be more expensive as they are recent houses, however older houses may be also expensive for historical reasoning. The area the house is in, the design, and the square feet living may affect the house more. 

*Price bin*

In [149]:
df['price_bin'] = df.apply(lambda x:'High' if x['price'] > 700000 else 'Medium' if x['price'] > 400000   else 'Low', axis = 1 )
df.head(2)

id       date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520 2014-10-13  221900.0         3       1.00         1180   
1  6414100192 2014-12-09  538000.0         3       2.25         2570   

   sqft_lot  floors  waterfront  view  ...  yr_renovated  zipcode      lat  \
0      5650     1.0           0     0  ...             0    98178  47.5112   
1      7242     2.0           0     0  ...          1991    98125  47.7210   

      long  sqft_living15  sqft_lot15  flag_bedroom  sqft_living2  \
0 -122.257           1340        5650             0           Low   
1 -122.319           1690        7639             0           Low   

   sqft_basement2  price_bin  
0               0        Low  
1               1     Medium  

[2 rows x 25 columns]

In [150]:
df['price_bin'].value_counts()

price_bin
Low       8839
Medium    8451
High      4323
Name: count, dtype: int64

In [158]:
df[['price_bin', 'sqft_living', 'grade', 'floors', 'view', 'waterfront', 'condition' ]].groupby('price_bin').agg({
     'sqft_living':'mean', 'grade': 'median', 
    'floors': 'median', 'view': 'mean', 'waterfront': 'mean', 'condition': 'median'
})

sqft_living  grade  floors      view  waterfront  condition
price_bin                                                             
High       3148.705991    9.0     2.0  0.703215    0.029378        3.0
Low        1578.665007    7.0     1.0  0.058038    0.001018        3.0
Medium     2057.412969    8.0     1.5  0.178795    0.003195        3.0

Through this thorough analysis i did, i realized that multiple factors contribute to house pricing much more than single factors alone. Eventhough the square footage of living space does have a huge impact, yet this impact is also influenced by other factors. Most of the attributes alone don't have the highest influence but when gathered the influence increased. The grade of the houses did have a high impact as well. Also, the presence of a good view and waterfront increased the prices exponentially, despite the fact that few houses have view/waterfront.  
To do such an analysis many cases have to be taken on all of the attributes, and they have to be crossed with many other attributes in order to find trends and patterns on the price.